internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -100,159 +102,204 @@ func (r *Reconciler) getNewMachineSet(ctx context.Context, d *clusterv1.MachineD
 		return nil, err
 	}
-	machineTemplateSpecHash := fmt.Sprintf("%d", hash)
-	newMSTemplate.Labels = mdutil.CloneAndAddLabel(d.Spec.Template.Labels,
-		clusterv1.MachineDeploymentUniqueLabel, machineTemplateSpecHash)
-
-	// Add machineTemplateHash label to selector.
-	newMSSelector := mdutil.CloneSelectorAndAddLabel(&d.Spec.Selector,
-		clusterv1.MachineDeploymentUniqueLabel, machineTemplateSpecHash)
-
-	minReadySeconds := int32(0)
-	if d.Spec.MinReadySeconds != nil {
-		minReadySeconds = *d.Spec.MinReadySeconds
-	}
-
-	// Create new MachineSet
-	newMS := clusterv1.MachineSet{
-		ObjectMeta: metav1.ObjectMeta{
-			// Make the name deterministic, to ensure idempotence
-			Name:      d.Name + "-" + apirand.SafeEncodeString(machineTemplateSpecHash),
-			Namespace: d.Namespace,
-			Labels:    make(map[string]string),
-			// Note: by setting the ownerRef on creation we signal to the MachineSet controller that this is not a stand-alone MachineSet.
-			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(d, machineDeploymentKind)},
-		},
-		Spec: clusterv1.MachineSetSpec{
-			ClusterName:     d.Spec.ClusterName,
-			Replicas:        new(int32),
-			MinReadySeconds: minReadySeconds,
-			Selector:        *newMSSelector,
-			Template:        newMSTemplate,
-		},
-	}
-
-	// Set the labels from newMSTemplate as top-level labels for the new MS.
-	// Note: We can't just set `newMSTemplate.Labels` directly and thus "share" the labels map between top-level and
+	log.V(4).Info("Created new MachineSet", "MachineSet", klog.KObj(newMS))
+	r.recorder.Eventf(d, corev1.EventTypeNormal, "SuccessfulCreate", "Created MachineSet %q", newMS.Name)
+
+	// Keep trying to get the MS. This will force the cache to update and prevent any future reconciliation of the
+	// MachineDeployment to reconcile with an outdated list of MachineSets which could lead to unwanted creation of a
+	// duplicate MachineSet.
+	if err := wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
+		ms := &clusterv1.MachineSet{}
+		if err := r.Client.Get(ctx, client.ObjectKeyFromObject(newMS), ms); err != nil {
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		return true, nil
+	}); err != nil {
+		return nil, errors.Wrapf(err, "failed to get the MachineSet %q after creation", klog.KObj(newMS))
+	}
+
+	err = r.updateMachineDeployment(ctx, d, func(innerDeployment *clusterv1.MachineDeployment) {
+		mdutil.SetDeploymentRevision(d, newRevision)
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to update revision annotation on MachineDeployment")
+	}
+
+	return newMS, nil
+}
+
+func (r *Reconciler) computeDesiredMachineSet(deployment *clusterv1.MachineDeployment, existingMS *clusterv1.MachineSet, oldMSs []*clusterv1.MachineSet, newRevision string) (*clusterv1.MachineSet, error) {
+	desiredMS := &clusterv1.MachineSet{}
+
+	desiredMS.SetGroupVersionKind(clusterv1.GroupVersion.WithKind("MachineSet"))
```

Comment:
Same here.

I would suggest defining the first few fields + empty maps for annotation/labels (top-level and in spec.template) as a struct. (also ownerref)

Should be easier to read and minimize the diff to before (and thus the things we have to compare to ensure we didn't break anything)